### PR TITLE
feat: adding CoW helper MVP

### PR DESCRIPTION
**Pull request by wei3erHase**
- [x] TODO: Vendor `GetTradeableOrder` from cow-amm after this PR gets merged: https://github.com/cowprotocol/cow-amm/pull/98
- [x] Helper MUST be reviewed by CoWSwap SC team

**BCoWHelper status**:

The BCoWHelper contract is designed to be view-called by solvers in order to fetch the relevant information that is needed to swap in a Pool, given a pricing vectore. The contract implements a CoW library, that supports only 2-token unweighted pools.

- Helper inherits `ICOWAMMPoolHelper` interface from cow-amm
- Helper implements `GetTradeableOrder` from cow-amm
- `tokens` method WILL revert with `PoolDoesNotExist` if:
    - Pool is not a children of the helper’s factory
    - Pool is not finalized (this one reverts with `BPool_PoolNotFinalized`)
    - **Pool has more than 2 tokens**
    - **Pool tokens have different weights**
- `order` method internally calls `tokens` (same reverting cases)
- `order` returns the commit pre-interaction and the order calculated in `GetTradeableOrder` library
- `factory` aims to the BCoWFactory and reads the AppData from it

### Considerations in limited compatibility:

- The unweighted nature of `GetTradeableOrder` is definite, the calculation doesn’t fall true for pools with different token weights, a new calculation needs to be done in order to support them (perhaps `GetTradeableWeightedOrder`?).
- The 2-token nature, on the other hand, is a workable limitation. The current design of the Helper just reverts if the Pool has more than 2 tokens. When we consider a 3-token pool (unweighted), we can imagine the interaction being: a 3-dimensional pricing vector is sent to the Helper, and the Helper will decide which of these 2 tokens should be traded (and build the `GPv2Order` + commit pre-interaction). This, at first sight is very difficult to implement, considering:
    - How do we create a 3-dimensional pricing vector?
    - How does the Helper know which tokens are best to trade?
    - How do we input a 3-dimensional array into the library?
    
    There is a workaround to this approach, and is making the helper accept 3-dimensional pricing vectors, but ONLY if 2 dimensions are non-zero. In that case, a `X/Y/Z` pool can be queried for `[X,Y,0]`, `[X,0,Z]`, and `[0,Y,Z]` separately, and reuse the 2-dimensional library that CoW provides.
    
- Pool level, there is no 2-token restriction for BCoWPools, Pools with up to 8 tokens can be created, and the BCoW flow supports swaps from any token to any other token in the Pool. When swapping in the Pool, the only relevant tokens are the token IN and OUT, no other token bounded to the pool has any effect on the swap, or in other way: **during swaps the Pool behaves always as 2-token**. So surplus measurements shouldn’t be affected by supporting multiple-token pools.

Summary:

- Helper is currently NOT supporting +2-token pools.
- Helper is currently NOT supporting weighted pools.
- Helper COULD support +2-token pools (iff the pricing vector has 2 non-zero values)
- Helper COULD NOT support weighted pools (requires new calcs).
- BCoWPool DOES support +2-token pools.
- BCoWPool DOES support weighted pools.

Comments:

- Adding a limitation to the BCoWFactory to allow only deployments of 2-token unweighted pools breaks the future compatibility of a newly developed Helper. In the event that the new calcs (to support weighted pools) are written, and some +2-token interaction is designed, the Factory would have to be redeployed in order to remove the check. And since supporting deployments of such Pools don’t cause any security issue, **we recommend that the BCoWFactory is deployed as is** (supporting +2-token weighted pools), that we keep the incompatibility checks at the Helper level, so that a new Helper can be added later without having to modify the Factory.

### CODE REVIEW

File: test/integration/BCoWHelper.t.sol [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+    GPv2Interaction.Data[] memory preInteractions;
+    GPv2Interaction.Data[] memory postInteractions;
+    bytes memory sig;
+    (ammOrder, preInteractions, postInteractions, sig) = helper.order(address(pool), prices);
+
+    // We expect a commit interaction in pre interactions
+    assertEq(preInteractions.length, 1);
+    assertEq(postInteractions.length, 0);
+
+    // Because of how we changed the price, we expect to buy DAI
```
Comment (fedgiac) [inline test/integration/BCoWHelper.t.sol:160]:
```suggestion
    // Because of how we changed the price, we expect to buy WETH
```

Leftover typo. :stuck_out_tongue:

File: test/integration/BCoWHelper.t.sol [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    basicPool.bind(address(WETH), VALID_AMOUNT, 4.2e18); // no weight
+
+    // finalize
+    weightedPool.finalize();
+    basicPool.finalize();
+
+    vm.stopPrank();
+  }
+
+  // NOTE: 1 ETH = 1000e6 DAI
```
Comment (fedgiac) [inline test/integration/BCoWHelper.t.sol:79]:
```suggestion
  // NOTE: 1 ETH = 1000 DAI
```

Typo? Or maybe I'm missing the point.

**Review by fedgiac — COMMENTED** [review]
Everything in the PR description is reasonable and the code is a good implementation for the helper (with one caveat, more on this later). The considerations/limitations are as expected. More in detail:

> we recommend that the BCoWFactory is deployed as is (supporting +2-token weighted pools), that we keep the incompatibility checks at the Helper level

The helper is indeed disposable and it's good to just deploy it as-is now, then replace it once we build up support for more tokens/weights (which is tricky for the reasons you mentioned).

About the caveat: the sell/buy price is computed with `GetTradeableOrder.getTradeableOrder`. This library works for a constant-product AMM like it's the case for the Balancer AMM as supported here. Still, the generated trade is (up to rounding) tight, in the sense that decreasing minimally the buy amount should cause the order to be invalid (manifested as an invalid signature, reverting with `BPool_TokenAmountOutBelowMinOut`). My concern is that, because of rounding, the order returned by `GetTradeableOrder.getTradeableOrder` might be slightly below the threshold for `calcOutGivenIn` and cause the helper to return a trade that isn't valid. The forked test right now works, but there could be variations of the input values (balances, weights (even if the same)) that may cause it to fail.
Concretely, this means that the CoW Protocol back-end would generate an invalid order and the AMM might be ignored for the batches where rounding causes the creation of a reverting trade.
A fix for this would be to replicate the `GetTradeableOrder` logic for the sell amount but use `calcOutGivenIn` for the sell amount. This should return the same result (up to rounding) but be unaffected by the issue above.

All other review comments are minor.

**Review by fedgiac — APPROVED** [review]
Under the assumption that #154 is merged, the code looks good for being used by the CoW Protocol back end.

File: test/unit/BCoWHelper.t.sol [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    factory.mock_call_isBPool(address(pool), true);
+    tokens[0] = makeAddr('token0');
+    tokens[1] = makeAddr('token1');
+    pool.set__tokens(tokens);
+    pool.set__records(tokens[0], IBPool.Record({bound: true, index: 0, denorm: VALID_WEIGHT}));
+    pool.set__records(tokens[1], IBPool.Record({bound: true, index: 1, denorm: VALID_WEIGHT}));
+    pool.set__totalWeight(2 * VALID_WEIGHT);
+    pool.set__finalized(true);
+
+    priceVector[0] = 1e18;
```
Comment (0xteddybear) [inline test/unit/BCoWHelper.t.sol:53]: I'd recommend using a price ratio that's not 1:1 so the unit tests can detect if we invert the order of parameters, for example

File: test/integration/BCoWHelper.t.sol [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+
+    // finalize
+    weightedPool.finalize();
+    basicPool.finalize();
+
+    vm.stopPrank();
+  }
+
+  // NOTE: 1 ETH = 1000 DAI
+  uint256 constant INITIAL_SPOT_PRICE = 0.001e18;
```
Comment (0xteddybear) [inline test/integration/BCoWHelper.t.sol:80]: feels weird to have a constant in between functions, are we sure it's best to disable solhint's `ordering` for tests?

File: test/integration/BCoWHelper.t.sol [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    } else {
+      assertEq(address(ammOrder.buyToken), address(WETH));
+    }
+
+    // it should not revert
+    pool.verify(ammOrder);
+  }
+
+  // NOTE: reverting test, weighted pools are not supported
+  function testWeightedOrder() public {
```
Comment (0xteddybear) [inline test/integration/BCoWHelper.t.sol:134]: I believe a unit test is enough for this

File: test/integration/BCoWHelper.t.sol [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+    assertEq(spotPrice, INITIAL_SPOT_PRICE);
+
+    _executeHelperOrder(pool, ammWethInitialBalance, ammDaiInitialBalance);
+
+    uint256 postSpotPrice = pool.getSpotPriceSansFee(address(WETH), address(DAI));
+    assertEq(postSpotPrice, 1_052_631_578_947_368);
+  }
+
+  // NOTE: this test checks that the order generated by the helper is valid
+  function testValidOrder(uint256 balanceToken0, uint256 balanceToken1, uint256 priceSkewness) public {
```
Comment (0xteddybear) [inline test/integration/BCoWHelper.t.sol:101]: I believe it would be best to have a hard-coded valid scenario (optionally with data as close to current blockchain state possible) in integration tests and perform fuzzing in unit tests, since they are executed with a higher `FOUNDRY_FUZZ_RUNS` parameter, and the contracts implementing the fuzzed logic are all offline-available libraries

File: test/integration/BCoWHelper.t.sol [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    (GPv2Order.Data memory ammOrder,,,) = helper.order(address(pool), prices);
+
+    if (priceSkewness > 10_000) {
+      assertEq(address(ammOrder.buyToken), address(DAI));
+    } else {
+      assertEq(address(ammOrder.buyToken), address(WETH));
+    }
+
+    // it should not revert
+    pool.verify(ammOrder);
```
Comment (0xteddybear) [inline test/integration/BCoWHelper.t.sol:130]: could we also test the signature with `isValidSignature`?

File: test/integration/BCoWHelper.t.sol [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+
+  BCoWFactory private ammFactory;
+  IBPool private weightedPool;
+  IBPool private basicPool;
+
+  IERC20 private constant DAI = IERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F);
+  IERC20 private constant USDC = IERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48);
+  IERC20 private constant WETH = IERC20(0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2);
+
+  uint256 constant VALID_AMOUNT = 1e6;
```
Comment (0xteddybear) [inline test/integration/BCoWHelper.t.sol:44]: this is a small amount, is it possible that you meant that to be used with USDC? or you want to cause very low slippage?

File: test/integration/BCoWHelper.t.sol [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+  address private vaultRelayer;
+
+  address private solver = 0x423cEc87f19F0778f549846e0801ee267a917935;
+
+  BCoWFactory private ammFactory;
+  IBPool private weightedPool;
+  IBPool private basicPool;
+
+  IERC20 private constant DAI = IERC20(0x6B175474E89094C44Da98b954EedeAC495271d0F);
+  IERC20 private constant USDC = IERC20(0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48);
```
Comment (0xteddybear) [inline test/integration/BCoWHelper.t.sol:41]: unused

**Review by 0xteddybear — COMMENTED** [review]
looking good, I'd move more checks onto the unit tests and clean up the integrations so the valid scenario they are executing is easier to follow

File: package.json [hunk 10]
```diff
@@ -42,6 +42,7 @@
     "@cowprotocol/contracts": "github:cowprotocol/contracts.git#a10f40788a",
     "@openzeppelin/contracts": "5.0.2",
     "composable-cow": "github:cowprotocol/composable-cow.git#24d556b",
+    "cow-amm": "github:cowprotocol/cow-amm.git#6566128",
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply package.json:45]: yes, they're not merging it to main, at least yet

File: src/contracts/BCoWHelper.sol [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+/**
+ * @title BCoWHelper
+ * @notice Helper contract that allows to trade on CoW Swap Protocol.
+ * @dev This contract supports only 2-token equal-weights pools.
+ */
+contract BCoWHelper is ICOWAMMPoolHelper, BMath {
+  using GPv2Order for GPv2Order.Data;
+
+  /// @notice The app data used by this helper's factory.
+  bytes32 public immutable APP_DATA;
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/contracts/BCoWHelper.sol:25]: hmmm we could make it internal tbh, this contract is supposed to be called only by the interface related methods, as it's al off-chain helper for solvers

File: src/contracts/BCoWHelper.sol [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  /// @inheritdoc ICOWAMMPoolHelper
+  function tokens(address pool) public view returns (address[] memory tokens_) {
+    // reverts in case pool is not deployed by the helper's factory
+    if (!IBCoWFactory(factory).isBPool(pool)) revert PoolDoesNotExist();
+    // call reverts with `BPool_PoolNotFinalized()` in case pool is not finalized
+    tokens_ = IBCoWPool(pool).getFinalTokens();
+    // reverts in case pool is not supported (non-2-token pool)
+    if (tokens_.length != 2) revert PoolDoesNotExist();
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply src/contracts/BCoWHelper.sol:113]: roger

File: test/unit/BCoWHelper.t.sol [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+
+    (
+      GPv2Order.Data memory order_,
+      GPv2Interaction.Data[] memory preInteractions,
+      GPv2Interaction.Data[] memory postInteractions,
+      bytes memory sig
+    ) = helper.order(address(pool), priceVector);
+
+    // it should return a valid pool order
+    assertEq(order_.receiver, GPv2Order.RECEIVER_SAME_AS_OWNER);
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/unit/BCoWHelper.t.sol:134]: no asserts on the amounts, they are recovered from the library which cannot be mocked, we're just testing here that what the library returns is, as for our standards, a valid order (also checked on integration test)

File: test/integration/BCoWHelper.t.sol (hunk 7 above)
Comment (wei3erHase) [reply test/integration/BCoWHelper.t.sol:130]: chose `verify` because `isValidSignature` actually does check everything at the same time, while verify only checks for the validity of the order

File: test/integration/BCoWHelper.t.sol (hunk 8 above)
Comment (wei3erHase) [reply test/integration/BCoWHelper.t.sol:44]: hmmm i believe this is the minimum amount out requestable, but i'll double check

File: test/integration/BCoWHelper.t.sol (hunk 5 above)
Comment (wei3erHase) [reply test/integration/BCoWHelper.t.sol:134]: this one was chosen to maintain the shape in case the feature gets added later, the unit test of the helper reverting is there, and the integration perspective we keep it here, in the future, another helper will not revert on weighted pools, and this test can be reimplemented not to revert

File: test/integration/BCoWHelper.t.sol (hunk 6 above)
Comment (wei3erHase) [reply test/integration/BCoWHelper.t.sol:101]: hmmm, i'm not entirely happy with fuzzed forked tests, but we do have little expectations, and we're trying to test a behaviour between multiple contracts: that what the helper returns is valid in the pool, so in a way yes, we could add the pool to the helper unit test contract and do it there, in another way i feel it could live within the integration scope, what do you think?

File: test/integration/BCoWHelper.t.sol (hunk 8 above)
Comment (wei3erHase) [reply test/integration/BCoWHelper.t.sol:44]: this is indeed the MIN_BALANCE to bind a token to

File: test/integration/BCoWHelper.t.sol (hunk 6 above)
Comment (wei3erHase) [reply test/integration/BCoWHelper.t.sol:101]: we're actually not using any oracle here either, it could be done all within a unit test scenario, but indeed requires more than 1 contract and it doesn't fit any tree

File: test/unit/BCoWHelper.tree [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+│   └── it should revert
+├── when pool has more than 2 tokens
+│   └── it should revert
+├── when pool tokens have different weights
+│   └── it should revert
+└── when pool is supported
+    └── it should return pool tokens
+
+BCoWHelperTest::order
+├── when the pool is not supported
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/unit/BCoWHelper.tree:19]: hmm, we cannot override a `public view` even if it's virtual :/
because we would need to do `address(this).call` within `tokens` to check it and that's not possible ona view method...

File: test/unit/BCoWHelper.t.sol [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+  MockBCoWFactory factory;
+  MockBCoWPool pool;
+  address invalidPool = makeAddr('invalidPool');
+  address[] tokens = new address[](2);
+  uint256[] priceVector = new uint256[](2);
+
+  uint256 constant VALID_WEIGHT = 1e18;
+
+  function setUp() external {
+    factory = new MockBCoWFactory(address(0), bytes32(0));
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/unit/BCoWHelper.t.sol:32]: i do think that deploying blank mocks and setting individual storages is preferrable, we need to note that:
- MockBCoWFactory doesn't deploy MockBCoWPools, lol
- we don't care what `solutionSettler` in BCoWFactory is set (zero is ok)
- we do care in BCoWPool (since is read on constructor)

File: test/unit/BCoWHelper.tree [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+│   └── it should revert
+├── when the pool is supported
+│   ├── it should query the domain separator from the pool
+│   ├── it should return a valid pool order
+│   ├── it should return a commit pre-interaction
+│   ├── it should return an empty post-interaction
+│   └── it should return a valid signature
+└── given a price vector
```
Comment (wei3erHase) [inline test/unit/BCoWHelper.tree:27]: cc-ing this to @drgorillamd, i'm breaking standard with `given` to not-branch but just make a test with a different focus, in this one in particular is fuzzing the balances and price vectors, it could have been done in the same scope as `when the pool is supported`, but think that separation of affairs is good there, since the code is looking at other things, wdyt?

File: test/integration/BCoWHelper.t.sol (hunk 5 above)
Comment (0xteddybear) [reply test/integration/BCoWHelper.t.sol:95]: ... if this is meant to stay then wdyt about removing the commented code below?

File: test/integration/BCoWHelper.t.sol [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+  function testBasicOrder() public {
+    IBCoWPool pool = IBCoWPool(address(basicPool));
+
+    uint256 spotPrice = pool.getSpotPriceSansFee(address(WETH), address(DAI));
+    assertEq(spotPrice, INITIAL_SPOT_PRICE);
+
+    _executeHelperOrder(pool);
+
+    uint256 postSpotPrice = pool.getSpotPriceSansFee(address(WETH), address(DAI));
+    assertEq(postSpotPrice, 1_052_631_578_947_368);
```
Comment (0xteddybear) [inline test/integration/BCoWHelper.t.sol:91]: I'd do a ranged assert with a computed amount, similar to how you did with `order.{buy,sell}Amount` below

File: test/manual-smock/MockBCoWHelper.sol [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  IBCoWFactory,
+  IBCoWPool,
+  ICOWAMMPoolHelper,
+  IERC20
+} from '../../src/contracts/BCoWHelper.sol';
+import {Test} from 'forge-std/Test.sol';
+
+contract MockBCoWHelper is BCoWHelper, Test {
+  // NOTE: manually added methods (internal immutable exposers not supported in smock)
+  function call__APP_DATA() external view returns (bytes32) {
```
Comment (0xteddybear) [inline test/manual-smock/MockBCoWHelper.sol:19]: why not make `_APP_DATA` public?

File: test/unit/BCoWHelper.t.sol [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+    bytes memory validSig = abi.encodePacked(pool, abi.encode(order_));
+    assertEq(keccak256(validSig), keccak256(sig));
+  }
+
+  function test_OrderGivenAPriceVector(uint256 priceSkewness, uint256 balanceToken0, uint256 balanceToken1) external {
+    // skew the price by max 50% (more could result in reverts bc of max swap ratio)
+    priceSkewness = bound(priceSkewness, 5000, 15_000);
+    vm.assume(priceSkewness != 10_000); // avoids no-skewness revert
+
+    balanceToken0 = bound(balanceToken0, 1e18, 1e36);
```
Comment (0xteddybear) [inline test/unit/BCoWHelper.t.sol:163]: is this to avoid underflows when dividing one by the other?

File: test/unit/BCoWHelper.t.sol [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+
+    // it should return an empty post-interaction
+    assertTrue(postInteractions.length == 0);
+
+    // it should return a valid signature
+    bytes memory validSig = abi.encodePacked(pool, abi.encode(order_));
+    assertEq(keccak256(validSig), keccak256(sig));
+  }
+
+  function test_OrderGivenAPriceVector(uint256 priceSkewness, uint256 balanceToken0, uint256 balanceToken1) external {
```
Comment (0xteddybear) [inline test/unit/BCoWHelper.t.sol:158]: I don't really follow _what_ are the `5000, `10_000` and `15_000`, much less _why_ we need them. Do you think it would be clearer if they had names? or some comments?

afaict we will _skew_ the price by a particular factor from current spot price in order to allow for naturally ocurring slippage, is that correct?

File: test/unit/BCoWHelper.t.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (0xteddybear) [reply test/unit/BCoWHelper.t.sol:135]: I think it's worth it to precompute some values to make sure we're not passing parameters to the library in the inverse order (or other silly-but-consequential bugs like that). This won't block merging since we're already doing it for the integrations, but is still desireable. (related to https://github.com/defi-wonderland/balancer-v1-amm/pull/121/files#r1678466826)

File: test/unit/BCoWHelper.t.sol [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+    vm.mockCall(tokens[1], abi.encodePacked(IERC20.balanceOf.selector), abi.encode(balanceToken1));
+
+    uint256[] memory prices = new uint256[](2);
+    prices[0] = balanceToken1;
+    prices[1] = balanceToken0 * priceSkewness / 10_000;
+
+    // it should return a valid pool order
+    (GPv2Order.Data memory ammOrder,,,) = helper.order(address(pool), prices);
+
+    assertEq(address(ammOrder.buyToken), priceSkewness > 10_000 ? tokens[0] : tokens[1]);
```
Comment (0xteddybear) [inline test/unit/BCoWHelper.t.sol:175]: are we sure these are not different branches?

File: test/manual-smock/MockBCoWHelper.sol (hunk 18 above)
Comment (wei3erHase) [reply test/manual-smock/MockBCoWHelper.sol:19]: is not required in the interface that CoWSwap provides `ICOWAMMPoolHelper`, they only request to expose `factory`, `tokens` and `order`

File: test/unit/BCoWHelper.t.sol (hunk 19 above)
Comment (wei3erHase) [reply test/unit/BCoWHelper.t.sol:163]: yasss, went for a valid but very broad range of numbers

File: test/unit/BCoWHelper.t.sol (hunk 3 above)
Comment (wei3erHase) [reply test/unit/BCoWHelper.t.sol:53]: i did! `1:1.05`

File: test/unit/BCoWHelper.t.sol (hunk 21 above)
Comment (wei3erHase) [reply test/unit/BCoWHelper.t.sol:175]: wdym by "not different branches"? 
i think is more of a different approach but same branch, `test_OrderWhenThePoolIsSupported` doesn't check the amounts and just use the default price vector (`1:1.05`), while `test_OrderGivenAPriceVector` fuzzes both balances and the skeweness but checks only that the pool can verify it, and that the token being sold is the correct one.

we could entirely fuzz `test_OrderWhenThePoolIsSupported`, but we'd be adding complexity to that test instead of maintaining the "valid scenario" approach, dunno wdyt of that

File: test/unit/BCoWHelper.t.sol (hunk 13 above)
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/unit/BCoWHelper.t.sol:135]: this is indeed tough because we don't have a precomputed amount that we're trading against the pool, but the helper is the one that says how much, we do have an expectation and is that at the end, the spot price should be similar to the pricing vector, will try that one

File: test/unit/BCoWHelper.tree (hunk 14 above)
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/unit/BCoWHelper.tree:19]: mistaken, we **can** override `tokens`, added an assertion that is being called 👍

[the remark this answers is not in the document]
Comment (0xteddybear) [reply test/unit/BCoWHelper.tree:19]: if that's the case then we can now get rid of this test, right? :eyes: I'd say testing 'when one of the lines of this function reverts, the revert bubbles up' is something we can delegate to the compiler :grin:

File: test/unit/BCoWHelper.t.sol [hunk 22]
```diff
@@ -167,12 +172,12 @@ contract BCoWHelperTest is Test {
...
     uint256[] memory prices = new uint256[](2);
     prices[0] = balanceToken1;
-    prices[1] = balanceToken0 * priceSkewness / 10_000;
+    prices[1] = balanceToken0 * priceSkewness / base;
 
     // it should return a valid pool order
     (GPv2Order.Data memory ammOrder,,,) = helper.order(address(pool), prices);
 
-    assertEq(address(ammOrder.buyToken), priceSkewness > 10_000 ? tokens[0] : tokens[1]);
+    assertEq(address(ammOrder.buyToken), priceSkewness > 1e18 ? tokens[0] : tokens[1]);
```
Comment (0xteddybear) [inline test/unit/BCoWHelper.t.sol:180]: use the `base` variable

File: src/contracts/BCoWHelper.sol [hunk 23]
```diff
@@ -0,0 +1,125 @@
...
+    bytes32 orderCommitment = order_.hash(domainSeparator);
+
+    preInteractions = new GPv2Interaction.Data[](1);
+    preInteractions[0] = GPv2Interaction.Data({
+      target: pool,
+      value: 0,
+      callData: abi.encodeWithSelector(IBCoWPool.commit.selector, orderCommitment)
+    });
+
+    return (order_, preInteractions, postInteractions, sig);
```
Comment (0xteddybear) [inline src/contracts/BCoWHelper.sol:103]: from what you laid out here: https://github.com/defi-wonderland/balancer-v1-amm/pull/156#issue-2405154590 I suppose we should skip the explicit return

File: test/unit/BCoWHelper.t.sol (hunk 21 above)
Comment (0xteddybear) [reply test/unit/BCoWHelper.t.sol]: I believe that to fully benefit from btt tests we should make the `.tree` fully match the branches available in the (code|domain). I've checked and they're effectively different branches: `node_modules/cow-amm/src/libraries/GetTradeableOrder.sol:56`

It was not obvious for me reviewing it, and it could have been easier to see had the information of `when token0 reserve * denominator > when token1 reserve * numerator` -> `buyToken is token1` been in the `.tree` file. I believe this is a case where you set some 'funny values' to get to that branch. I consider this PR should include:

- [x] fully fuzzed tests to report on rounding errors (this one is ok for these purposes, we could also be purists and have one fully-fuzzed for each branch. Definetly not required.) This doesnt fit neatly into the .tree but I can live with that.
- [ ] 'valid scenario' 'funny value'd unit tests that are described in .tree and walk every available branch. In this case I'd consider the library an implementaiton detail and walk their branches as well, as that's critical to ensure this helper works as intended. This includes:
    - [ ] one testcase forcing token0 to be buyToken
    - [ ] one testcase forcing token1 to be buyToken
    - [ ] Some assertions on `src/contracts/BCoWHelper.sol:74`, since that defines two branches as well. Since they are coupled to the above, I don't know if they'll be their own separate two testcases or extra asserts inside them, do whatever feels right.

tl;dr: fuzzing is ok as is, we should have tests manually walking each branch.

**Review by 0xteddybear — COMMENTED** [review]
having btt tests walk all branches is highly desireable, the rest are nitpicks.

File: test/unit/BCoWHelper.tree (hunk 16 above)
Comment (drgorillamd) [reply test/unit/BCoWHelper.tree:28]: `given` ftw :D cc @gas1cent ?

Comment (gas1cent) [reply test/unit/BCoWHelper.tree]: OK FINE

**Review by 0xteddybear — APPROVED** [review]
having fuzzing cover separate code branches in different tests is a good compromise, however as an improvement I would have preferred:

- for the unit tests to include `when-` branches with 'funny values' that cause the different branches' execution with no fuzzing setup.
- the above to include asserts on `{buy,sell}Amount` (not a hard requirement since they are covered in the integrations)
- fuzzing separate from that, in `given` branches, possibly with less assertions

File: src/contracts/BCoWHelper.sol (hunk 23 above)
Comment (wei3erHase) [reply src/contracts/BCoWHelper.sol:103]: i think is worth calling it "non-obvious return", specially since `postInteractions` will be empty and we cannot not declare it (compiler warnings)

File: test/unit/BCoWHelper.t.sol [hunk 24]
```diff
@@ -159,26 +160,62 @@ contract BCoWHelperTest is Test {
...
     // skew the price by max 50% (more could result in reverts bc of max swap ratio)
-    uint256 base = 1e18;
-    priceSkewness = bound(priceSkewness, 0.5e18, 1.5e18);
-    vm.assume(priceSkewness != base); // avoids no-skewness revert
+    // avoids no-skewness revert
+    priceSkewness = bound(priceSkewness, BASE + 0.0001e18, 1.5e18);
 
-    balanceToken0 = bound(balanceToken0, 1e18, 1e36);
-    balanceToken1 = bound(balanceToken1, 1e18, 1e36);
+    balanceToken0 = bound(balanceToken0, 1e18, 1e27);
```
[the remark this answers is not in the document]
Comment (wei3erHase) [reply test/unit/BCoWHelper.t.sol:172]: avoiding overflows
